LANDSCAPE/COMPARISONS/ARGOCD-FLUX/index.md

```diff
@@ -0,0 +1,275 @@
+
+With Argo CD you have more control over ordering of individual resources (i.e. specific YAML file) with [Syncwaves and Phases](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/). This allows you to apply each YAML manifest one at a time. Using Syncwaves and Phases, Argo CD will wait until one YAML is applied (and healthy) before moving onto applying the next one. 
+
+There is a controller that supplements Argo CD’s Application controller called [Application Sets](https://github.com/argoproj-labs/applicationset). This add-on lets users deploy many Applications across one or many managed clusters.
+
+
+# Secrets Management
+
+GitOps practice of storing the source of truth in git, has had some contention with respect to storing [Kubernetes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/). This has to do with the fact that Secrets often contain sensitive 	information like passwords or tokens, and these secrets are only encoded.
+
+Both Flux and Argo CD have mechanisms in place to handle the encrypting of secrets.
+
+
+## FluxCD
+
+FluxCD provides a guide for managing encrypted secrets with [Mozilla SOPS](https://fluxcd.io/docs/guides/mozilla-sops/) or [Sealed Secrets](https://fluxcd.io/docs/guides/sealed-secrets/) so that Kubernetes secrets can be stored safely in Git, secured by any of OpenPGP, AWS KMS, GCP KMS and Azure Key Vault. Flux’s Kustomization objects natively support decrypting Mozilla SOPS secrets.
+
+
+## Argo CD
+
+Argo CD can also [integrate with a variety of Secret management tools](https://argo-cd.readthedocs.io/en/stable/operator-manual/secret-management/) to store these secrets in Git, safely. It leaves this choice up to the user. Some users have taken advantage of Argo CD’s plugin feature to create plugins, for example, the[ Argo CD Vault Plugin](https://github.com/IBM/argocd-vault-plugin). This “hands off” approach just treats secrets as another _Application_ in the cluster.
+
+
+# Webhook receivers
+
+Webhooks can be used to help with immediate reconciliation with traditional, event driven, workflows. This doesn’t take the place of the reconciliation loop, but rather acts as an “on demand” reconciliation if you don’t want to wait after making a commit to a repo.
+
+The two tools have similar approaches so we just highlight where to get more information.
+
+
+## FluxCD
+
+FluxCD includes support for incoming webhooks that make GitOps pull-based pipelines as fast and responsive as push-based pipelines. [Webhook receivers](https://fluxcd.io/docs/guides/webhook-receivers/) can notify the Flux controllers about changes in Git or Helm repositories, or other sources like Image Repositories.
+
+
+## Argo CD
+
+Argo CD also has a mechanism to receive [Webhooks](https://argo-cd.readthedocs.io/en/stable/operator-manual/webhook/) to shorten the time it takes to check for updates to a Git repository. Normally this is done every 3 minutes.
+
+
+# Alerting and notifications
```

Comment:
```suggestion
# Rollbacks

• Manual `git revert` and `git revert -m` (for merge commits) appear to be the recommended approach to rollback in Git.  Neither Flux nor Argo CD execute the `git` commands on behalf of humans.

## Flux

Flux Helm Controller supports rollback, but it may require manual user intervention to reconcile broken release states.  Flux Kustomize Controller has no mention of rollbacks.

## Argo CD

**Argo CD supports rollback as an emergency measure.  There are issues with pruning and the lack of hooks.


# Alerting and notifications
```
‎

@christianh814 et al.

I searched for rollbacks and here is what I found:

• Manual `git revert` and `git revert -m` (for merge commits) appear to be the recommended approach to rollback in Git.  Neither Flux nor Argo CD execute the `git` commands on behalf of humans.

• **Argo CD** supports rollback as an emergency measure.  There are issues with pruning and the lack of hooks.

• **Flux v2 Helm Controller** supports rollback, but it may require manual user intervention to reconcile broken release states.

• **Flux v2 Kustomize Controller** has no mention of rollbacks.

• **Flux v1 Helm Operator** supports rollbacks, but it may require manual user intervention to recover.

• **Flux v1 (Kustomize) Operator** doesn't support rollbacks.

• **Kubernetes** supports
```kubectl rollout undo deployment/[app]-deployment --to-revision=#```
but it would misalign with GitOps tools.

• **GitOps Working Group Slack** has a conversation about an idea of storing a release artifact from every change applied to production for rollbacks, instead of running `git`.

---

**References below:**

---

**Argo CD**:

From https://argo-cd.readthedocs.io/en/stable/user-guide/commands/argocd_app_rollback/

> `argocd app rollback APPNAME [ID] [flags]`

From https://github.com/argoproj/argo-cd/issues/1927

> When you rollback you cannot:
> 1. Decide to prune resources
> 2. Run hooks

From https://codefresh.io/docs/docs/ci-cd-guides/gitops-deployments/#rolling-back-git-versions

> In the GitOps dashboard you will also see a complete history of all past deployments as recorded in Git. You can select any of the previous versions and rollback your application to the respective version.
>
> ![rollback.png](https://codefresh.io/docs/images/guides/gitops/rollback.png)
> _Rolling back to a previous version_
>
> The Rollback simply informs the cluster to use a different git hash for the sync process. It doesn’t affect your Git repository and ArgoCD will now show your application as out-of-sync (because the last Git commit no longer matches the status of the cluster).
>
> This rollback behavior is best used as an emergency measure after a failed deployment where you want to bring the cluster back to a previous state in a temporary manner. If you wish to keep the current rollback status as a permanent status it is best to use the standard `git reset/revert` commands and change the GitOps repository to its desired state.

---

**Flux v2**:

From https://fluxcd.io/docs/components/helm/helmreleases/

> The remediation action for upgrade failures is by default a rollback

From https://github.com/fluxcd/flux2/issues/1882#issuecomment-932209870

> Flux can correct drift in-cluster, such as a manual helm upgrade that is rollback at the specified interval.

From https://github.com/fluxcd/flux2/issues/1878#issuecomment-962208864

> helmrelease "upgrade retries exhausted" regression
>
> The workaround for now is this:
```flux suspend hr <release_name>```
```flux resume hr <release_name>```
> it will reconcile broken release states such as "exhausted" and "another rollback/release is in progress". 

---

**Flux v1**:

From https://github.com/fluxcd/flux/issues/3201#issuecomment-674854822

> Flux does not rollback nor do we consider adding such a feature as the git state can diverge between sync and apply.

From https://github.com/fluxcd/flux/issues/2132

> Failed Helm releases require manual user intervention to recover
>
> There already is an issue for rollbacks: [#1960](https://github.com/fluxcd/flux/issues/1960) (being worked on in [#2006](https://github.com/fluxcd/flux/pull/2006)), including an explanation on why we can not enable it by default for every HelmRelease.

From https://github.com/fluxcd/flux/issues/1960#issuecomment-488664697

> One of the challenges I am facing is the fact that in case an upgrade fails and it rolls back the release, it will end up in an infinite loop of retrying the upgrade (and rolling back) until a human notices and fixes the error.
>
> This is harmless in theory, it will however result in about 40 nonsense revisions/hour with the default interval settings.

From https://github.com/fluxcd/flux/issues/2221

> **Describe the feature**
>
> A `--previous` flag for `fluxctl release`, which rolls back the given `--workload` to the image before the current image (with filters being applied if configured).
>
> This is useful in more complex CI setups where for example an e2e smoke test can detect (and with this feature easily roll back) a faulty release.
>
> **NB:** for automated resources we would probably need to lock it at the same time, as otherwise it will roll forward again (`--lock`?).
>
> **What would the new user story look like?**
>
> 1. CI performs a release with `fluxctl release --workload <namespace>:<type>/<name>`
> 2. e2e smoke test detects abnormalities, CI now wants to perform a rollback
> 3. CI runs `fluxctl release --workload <namespace>:<type>/<name> --previous`
> 4. resource is now one version behind 'latest'

From https://github.com/fluxcd/flux/issues/2221#issuecomment-626792296

> Maybe this can be solved using kubectl rollout undo, or am I missing something?
>
> UPDATE:
> The problem is that we would break GitOps, right?

---

**Kubernetes**:

From https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#rolling-back-to-a-previous-revision

> **Rolling Back to a Previous Revision**
>
> Now you've decided to undo the current rollout and rollback to the previous revision:
> `kubectl rollout undo deployment/nginx-deployment`
>
> Alternatively, you can rollback to a specific revision by specifying it with --to-revision:
> `kubectl rollout undo deployment/nginx-deployment --to-revision=2`

---

**GitOps Working Group Slack**:

From https://cloud-native.slack.com/archives/C01G9DEE88M/p1637687831177600

> How hard would it be for GitOps controllers/operators/etc. to store a release artifact from every change applied to prod. That way your rollbacks are potentially easier than mucking with git itself. Food for thought.

---

**Git**:

From https://opensource.com/article/18/6/git-reset-revert-rebase-commands

> **Revert or reset?**
>
> Why would you choose to do a revert over a reset operation? If you have already pushed your chain of commits to the remote repository (where others may have pulled your code and started working with it), a revert is a nicer way to cancel out changes for them. This is because the Git workflow works well for picking up additional commits at the end of a branch, but it can be challenging if a set of commits is no longer seen in the chain when someone resets the branch pointer back.
>
> This brings us to one of the fundamental rules when working with Git in this manner: Making these kinds of changes in your local repository to code you haven't pushed yet is fine. But avoid making changes that rewrite history if the commits have already been pushed to the remote repository and others may be working with them.

From https://stackoverflow.com/a/7100005

> **How to revert a merge commit that's already pushed to a remote branch?**
>
> The -m option specifies the parent number. This is because a merge commit has more than one parent, and Git does not know automatically which parent was the mainline, and which parent was the branch you want to un-merge.
>
> In this situation, git revert 8f937c6 -m 1 will get you the tree as it was in 8989ee0, and git revert -m 2 will reinstate the tree as it was in 7c6b236.

---

From https://codefresh.io/about-gitops/pains-gitops-1-0/

> **There is no standard practice for GitOps rollbacks**
>
> The fact that all your cluster history is in Git makes rollbacks in GitOps (supposedly) very easy.
>
> If you want to rollback to a previous version, you simply use a past commit for your sync operation.
>
> In practice, however, it is not entirely clear what exactly we mean by “use a previous commit”. Different people use different ways to rollback:
>
> 1. You can simply point your cluster to a previous Git hash and let your GitOps tool sync that hash. This is the fastest way to rollback, but by definition leaves your cluster in an inconsistent state as the cluster does not contain what is described in the last Git commit.
>
> 2. You can use the standard Git reset, Git revert commands, and again let your GitOps tool perform the sync operation as usual. This keeps the GitOps promise (of having in the cluster what is in the Git repo) but of course, requires manual intervention.
>
> 3. You can have a combination where the GitOps tool itself both syncs a previous Git hash to the cluster and auto-commits (or reverts) to the git repo in order to keep the consistency. This is very complex to accomplish and not all teams want to let their deployment solution have write access to their Git repo.
>
> It goes without saying that different people might want a completely different approach for rollbacks. At the time of writing, however, the present GitOps tools have very little support and guidance on how you perform a rollback in a standard way.